dialogflow/detect_intent_stream.rb

```diff
@@ -0,0 +1,132 @@
+# Copyright 2018 Google, Inc
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+require "securerandom"
+require "forwardable"
+
+class EnumeratorQueue
+  extend Forwardable
+  def_delegators :@q, :push
+
+  # @private
+  def initialize sentinel
+    @q = Queue.new
+    @sentinel = sentinel
+  end
+
+  # @private
+  def each_item
+    return enum_for(:each_item) unless block_given?
+    loop do
+      r = @q.pop
+      break if r.equal? @sentinel
+      fail r if r.is_a? Exception
+      yield r
+    end
+  end
+end
+
+def detect_intent_stream project_id:, session_id:, audio_file_path:,
+                        language_code:
+  # [START dialogflow_detect_intent_stream]
+  # project_id = "Your Google Cloud project ID"
+  # session_id = "mysession"
+  # audio_file_path = "resources/book_a_room.wav"
+  # language_code = "en-US"
+  
+  require "google/cloud/dialogflow"
+  require "monitor"
+
+  session_client = Google::Cloud::Dialogflow::Sessions.new
+  session = session_client.class.session_path project_id, session_id
+  puts "Session path: #{session}"
+
+  audio_config = { 
+    audio_encoding: :AUDIO_ENCODING_LINEAR_16,
+    sample_rate_hertz: 16000,
+    language_code: language_code
+  }
+  query_input = { audio_config: audio_config }
+  streaming_config = { session: session, query_input: query_input }
+
+  # To signal the main thread when all responses have been processed
+  completed = false
+
+  # Use session_client as the sentinel to signal the end of queue
+  request_queue  = EnumeratorQueue.new(session_client)
+
+  # The first request needs to be the configuration.
+  request_queue.push(streaming_config)
+
+  # Consume the queue and process responses in a separate thread
+  Thread.new do
+    session_client.streaming_detect_intent(request_queue.each_item).each do |response|
+      if response.recognition_result
+        puts "Intermediate transcript: #{response.recognition_result.transcript}\n"
```

Comment:
out of curiosity, does intermediate transcript print output right now?

Reply:
it does:
```
% ruby detect_intent_stream.rb resources/book_a_room.wav
Session path: projects/rising-sea-112358/agent/sessions/80f3f76f-3146-48a4-9554-a10727e6da7f
Intermediate transcript: book
Intermediate transcript: buca
Intermediate transcript: book of
Intermediate transcript: book a room
Intermediate transcript: book a room
Query text:        book a room
Intent detected:   room.reservation
Intent confidence: 1.0
Fulfillment text:  I can help with that. Where would you like to reserve a room?
```